open_flamingo/src/factory.py

```diff
@@ -27,99 +32,122 @@ def create_model_and_transforms(
     Args:
         clip_vision_encoder_path (str): path to pretrained clip model (e.g. "ViT-B-32")
         clip_vision_encoder_pretrained (str): name of pretraining dataset for clip model (e.g. "laion2b_s32b_b79k")
-        lang_encoder_path (str): path to pretrained language encoder
+        lang_model_path (str): path to pretrained language encoder
         tokenizer_path (str): path to pretrained tokenizer
         cross_attn_every_n_layers (int, optional): determines how often to add a cross-attention layer. Defaults to 1.
         use_local_files (bool, optional): whether to use local files. Defaults to False.
         decoder_layers_attr_name (str, optional): name of the decoder layers attribute. Defaults to None.
-        freeze_lm_embeddings (bool, optional): whether to freeze LM input embeddings when configuring Perceiver.
         cache_dir (str, optional): path to cache directory for downloading OpenClip/HF weights.
+        gradient_checkpointing (bool, optional): whether to use gradient checkpointing. Defaults to False.
+        verbose (bool, optional): whether to print model info. Defaults to True.
     Returns:
         Flamingo: Flamingo model from pretrained vision and language encoders
         Image processor: Pipeline to preprocess input images
         Tokenizer: A tokenizer for the language model
     """
+
+    assert model_family in SUPPORTED_MODEL_FAMILIES
+
+    # load vision encoder
     vision_encoder, _, image_processor = open_clip.create_model_and_transforms(
         clip_vision_encoder_path,
         pretrained=clip_vision_encoder_pretrained,
         cache_dir=cache_dir,
     )
-    # set the vision encoder to output the visual features
     vision_encoder.visual.output_tokens = True
+    vision_encoder = vision_encoder.visual
+    vis_hidden_dim = open_clip.get_model_config(clip_vision_encoder_path)["vision_cfg"][
+        "width"
+    ]
 
+    # load tokenizer and ensure there is a pad token
     text_tokenizer = AutoTokenizer.from_pretrained(
         tokenizer_path,
         local_files_only=use_local_files,
         trust_remote_code=True,
         cache_dir=cache_dir,
     )
-    # add Flamingo special tokens to the tokenizer
-    text_tokenizer.add_special_tokens(
-        {"additional_special_tokens": ["<|endofchunk|>", "<image>"]}
-    )
     if text_tokenizer.pad_token is None:
-        # Issue: GPT models don't have a pad token, which we use to
-        # modify labels for the loss.
-        text_tokenizer.add_special_tokens({"pad_token": "<PAD>"})
```

Comment:
Maybe we should keep the pad token as <PAD> so it is easier to be compatible with older weights. Thinking we can just reformat the embedding of those to be in the decoupled setup.

Reply:
Could we remove inserted pad embeddings from the old weights? (If we need to reformat the old weights anyway)

Reply:
sounds good